### PR TITLE
feature: logic of login by email and by google

### PR DESCRIPTION
**Pull request by ivanderkachov**
Logic of Login by email and Google are provided
Two endpoints are added to AuthModule:
- /login - login by email and password
- /google/login - login by Google (verifying of access_token)
Also added logic of excluding password in server response

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -15,6 +15,11 @@ async function start() {
   const document = SwaggerModule.createDocument(app, config);
   SwaggerModule.setup('api', app, document);
   app.useGlobalPipes(new ValidationPipe());
+  app.enableCors({
+    allowedHeaders: ['content-type'],
+    origin: 'http://localhost:3000',
```
Comment (lyudad) [inline src/main.ts]: all links should be in env file. today yo have localhost tomorrow https://mysupercoolwesite.com

File: src/modules/auth/dtos/loginUser.dto.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { Exclude } from 'class-transformer';
+import { IsEmail, IsString } from 'class-validator';
+
+export class LoginUserDto {
+  @IsString()
+  @IsEmail()
+  readonly email: string;
```
Comment (lyudad) [inline src/modules/auth/dtos/loginUser.dto.ts:7]: look like eslint not work. there should be empty line

File: src/modules/auth/services/auth.service.ts [hunk 3]
```diff
@@ -55,4 +65,41 @@ export class AuthService implements IAuthService {
       throw new HttpException(`${error}`, HttpStatus.INTERNAL_SERVER_ERROR);
     }
   }
+
+  async login(userDto: CreateUserDto) {
```
Comment (lyudad) [inline src/modules/auth/services/auth.service.ts]: where are return types. check git checklist!

File: src/modules/auth/services/auth.service.ts [hunk 4]
```diff
@@ -55,4 +65,41 @@ export class AuthService implements IAuthService {
...
+          return new LoginUserDto(user);
+        }
+      }
+      throw new UnauthorizedException(`Authorization error`);
+    } catch (error) {
+      throw new HttpException(`${error}`, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  async loginByGoogle(token: string) {
```
Comment (lyudad) [inline src/modules/auth/services/auth.service.ts]: same

File: src/modules/auth/services/auth.service.ts [hunk 5]
```diff
@@ -55,4 +65,41 @@ export class AuthService implements IAuthService {
...
+    try {
+      const client = new OAuth2Client(
+        process.env.GOOGLE_CLIENT_ID,
```
Comment (lyudad) [inline src/modules/auth/services/auth.service.ts]: read nest documentation how to use envs

**Comment by ivanderkachov** [comment]
I fixed all according to your comments: add variables to env file, add eslint rule, add return types

File: src/main.ts (hunk 1 above)
Comment (ivanderkachov) [reply src/main.ts]: Updated

File: src/modules/auth/dtos/loginUser.dto.ts (hunk 2 above)
Comment (ivanderkachov) [reply src/modules/auth/dtos/loginUser.dto.ts:7]: Updated

File: src/modules/auth/services/auth.service.ts (hunk 3 above)
Comment (ivanderkachov) [reply src/modules/auth/services/auth.service.ts]: Updated

File: src/modules/auth/services/auth.service.ts (hunk 4 above)
Comment (ivanderkachov) [reply src/modules/auth/services/auth.service.ts]: Updated

File: src/modules/auth/services/auth.service.ts (hunk 5 above)
Comment (ivanderkachov) [reply src/modules/auth/services/auth.service.ts]: Updated

File: src/main.ts [hunk 6]
```diff
@@ -15,6 +15,11 @@ async function start() {
   const document = SwaggerModule.createDocument(app, config);
   SwaggerModule.setup('api', app, document);
   app.useGlobalPipes(new ValidationPipe());
+  app.enableCors({
+    allowedHeaders: ['content-type'],
+    origin: process.env.CLIENT_HOST,
+    credentials: true,
+  });
   await app.listen(PORT, () => console.log(`Server started on port ${PORT}`));
```
Comment (lyudad) [inline src/main.ts]: Снимок экрана 2022-11-15 в 20 29 24
your eslint rules not configured right, this rule is missed

Comment (lyudad) [reply src/main.ts]: also where is try/catch you are inside async function?

Comment (ivanderkachov) [reply src/main.ts]: updated

File: src/modules/auth/controllers/auth.controller.ts [hunk 7]
```diff
@@ -40,4 +44,22 @@ export class AuthController {
   confirm(@Query(ValidationPipe) query: ConfirmAccountDto) {
     return this.authService.confirmEmail(query.id);
   }
+
+  @ApiBody({ type: CreateUserDto })
+  @ApiResponse({ status: 200, description: 'Login user' })
+  @Post(Routes.LOGIN)
+  @UsePipes(ValidationPipe)
+  @UseInterceptors(ClassSerializerInterceptor)
+  login(@Body() userDto: CreateUserDto) {
```
Comment (lyudad) [inline src/modules/auth/controllers/auth.controller.ts]: Снимок экрана 2022-11-16 в 22 33 47

File: src/modules/auth/controllers/auth.controller.ts [hunk 8]
```diff
@@ -40,4 +44,22 @@ export class AuthController {
...
+    return this.authService.login(userDto);
+  }
+
+  @ApiBody({ type: TokenDto })
+  @ApiResponse({ status: 200, description: 'Login user' })
+  @Post(Routes.GOOGLE_LOGIN)
+  @UsePipes(ValidationPipe)
+  @UseInterceptors(ClassSerializerInterceptor)
+  loginByGoogle(@Body('token') token: string) {
```
Comment (lyudad) [inline src/modules/auth/controllers/auth.controller.ts]: Снимок экрана 2022-11-16 в 22 33 47

File: src/modules/auth/services/auth.service.ts [hunk 9]
```diff
@@ -7,13 +13,18 @@ import { InjectRepository } from '@nestjs/typeorm';
 import { User } from 'src/database/entities/users.entity';
 import { Repository } from 'typeorm';
 import { MailService } from 'src/modules/mail/services/mail.service';
+import { comparePassword } from 'src/utils/hash';
+import { LoginUserDto } from '../dtos/loginUser.dto';
+import { OAuth2Client } from 'google-auth-library';
+import { ConfigService } from '@nestjs/config';
```
Comment (lyudad) [inline src/modules/auth/services/auth.service.ts:19]: Снимок экрана 2022-11-16 в 22 35 03

File: src/modules/auth/controllers/auth.controller.ts (hunk 7 above)
Comment (ivanderkachov) [reply src/modules/auth/controllers/auth.controller.ts]: Updated

File: src/modules/auth/controllers/auth.controller.ts (hunk 8 above)
Comment (ivanderkachov) [reply src/modules/auth/controllers/auth.controller.ts]: Updated

File: src/modules/auth/services/auth.service.ts (hunk 9 above)
Comment (ivanderkachov) [reply src/modules/auth/services/auth.service.ts:19]: Updated

File: src/utils/hash.ts [hunk 10]
```diff
@@ -2,5 +2,12 @@ import * as bcrypt from 'bcrypt';
 
 export async function hashPassword(password: string) {
   const salt = await bcrypt.genSalt();
+
   return bcrypt.hash(password, salt);
 }
+
+export async function comparePassword(hashPassword: string, password: string) {
```
Comment (lyudad) [inline src/utils/hash.ts]: Снимок экрана 2022-11-20 в 20 05 06

Comment (ivanderkachov) [reply src/utils/hash.ts]: Updated

**Pull request closed without merging**